### PR TITLE
Linking: replace conflicting pointer types when one declaration is extern

### PR DESCRIPTION
**Pull request by tautschnig**

### CODE REVIEW

**Review by smowton — APPROVED** [review]
lgtm

**Review by martin-cs — APPROVED** [review]
The parts of it that touch my world look good.

File: src/ansi-c/c_typecheck_base.cpp [hunk 1]
```diff
@@ -67,6 +67,14 @@ void c_typecheck_baset::typecheck_symbol(symbolt &symbol)
...
     symbol.is_static_lifetime=true;
+
+    if(symbol.value.is_not_nil())
+    {
+      // according to the C standard this should be an error, but at least some
+      // versions of Visual Studio insist to use this in their C library
+      warning().source_location = symbol.value.find_source_location();
+      warning() << "`extern' symbol should not have an initializer" << eom;
+    }
   }
```
Comment (kroening) [inline src/ansi-c/c_typecheck_base.cpp:79]: Turn into error when in gcc or clang mode?

File: src/util/replace_symbol.cpp [hunk 2]
```diff
@@ -29,7 +29,8 @@ void replace_symbolt::insert(
 
 bool replace_symbolt::replace(
   exprt &dest,
-  const bool replace_with_const) const
+  const bool replace_with_const,
+  const bool ignore_type_change) const
 {
```
Comment (kroening) [inline src/util/replace_symbol.cpp]: I think this is getting to bizarre, and to linker specific. I'd make that a separate piece of code, just for the linker.

Comment (tautschnig) [reply src/util/replace_symbol.cpp]: Work in progress, #2720 and #2721 are the first PRs to enable this. More to come.

**Comment by tautschnig** [comment]
This is now implemented on top of #2723 with the suggested, linker-specific implementation.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 6e921c207800a812bbf421d62688e4948960f3c0).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 6e921c207800a812bbf421d62688e4948960f3c0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/83986092
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: c72055567e2b126fcf756af69605ddc33cce2809).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84011100

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: e8b03a7f39ec0b4bcca66a55c36ae9ffd8c2e8c2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84094002

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: d7dae290ca9ed8fba033c3cfe75d9e60e322913c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86154692

**Comment by tautschnig** [comment]
@kroening ping?

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 2f60e132b9b4ae989ca2301d996ab8126995ca9e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86766275

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 80ff4e58a85d98a487e765a1966a342bd68265ab).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86776976

**Review by kroening — APPROVED** [review]

**Pull request merged**